File: app/src/main/java/com/cyl/musiclake/ui/music/search/SearchActivity.java
package com.cyl.musiclake.ui.music.search;

import android.support.annotation.NonNull;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.animation.DecelerateInterpolator;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;

import com.cyl.musiclake.R;
import com.cyl.musiclake.base.BaseActivity;
import com.cyl.musiclake.bean.HotSearchBean;
import com.cyl.musiclake.bean.Music;
import com.cyl.musiclake.bean.SearchHistoryBean;
import com.cyl.musiclake.common.Constants;
import com.cyl.musiclake.common.NavigationHelper;
import com.cyl.musiclake.data.db.DaoLitepal;
import com.cyl.musiclake.player.PlayManager;
import com.cyl.musiclake.ui.music.dialog.BottomDialogFragment;
import com.cyl.musiclake.ui.music.local.adapter.SongAdapter;
import com.cyl.musiclake.utils.AnimationUtils;
import com.cyl.musiclake.utils.LogUtil;
import com.google.android.flexbox.FlexDirection;
import com.google.android.flexbox.FlexboxLayoutManager;
import com.google.android.flexbox.JustifyContent;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.OnClick;

/**
 * 作者：yonglong on 2016/9/15 12:32
 * 邮箱：643872807@qq.com
 * 版本：2.5
 */
public class SearchActivity extends BaseActivity<SearchPresenter> implements SearchContract.View {

    private static final String TAG = "SearchActivity";
    //搜索信息
    private String queryString;
    private SongAdapter mAdapter;
    private SearchHistoryAdapter historyAdapter;
    private HotSearchAdapter hotSearchAdapter;

    @BindView(R.id.recyclerView)
    RecyclerView resultListRcv;
    @BindView(R.id.suggestions_list)
    RecyclerView historyRcv;
    @BindView(R.id.hotSearchView)
    View hotSearchView;
    @BindView(R.id.hotSearchRcv)
    RecyclerView hotSearchRcv;
    @BindView(R.id.history_panel)
    View historyPanel;
    @BindView(R.id.toolbar_search_edit_text)
    EditText searchEditText;
    @BindView(R.id.toolbar_search_container)
    View searchToolbarContainer;

    @OnClick(R.id.clearAllIv)
    void clearAll() {
        DaoLitepal.INSTANCE.clearAllSearch();
        historyAdapter.setNewData(null);
    }

    @OnClick(R.id.toolbar_search_clear)
    void clearQuery() {
        queryString = "";
        searchEditText.setText("");
    }

    private List<Music> searchResults = new ArrayList<>();
    private List<SearchHistoryBean> suggestions = new ArrayList<>();

    private int mCurrentCounter = 10;
    private int limit = 10;
    private int mOffset = 0;


    @Override
    protected int getLayoutResID() {
        return R.layout.acitvity_search;
    }

    @Override
    protected void initView() {
        showSearchOnStart();
    }

    private void showSearchOnStart() {
        searchEditText.setText(queryString);

        if (TextUtils.isEmpty(queryString) || TextUtils.isEmpty(searchEditText.getText())) {
            searchToolbarContainer.setTranslationX(100);
            searchToolbarContainer.setAlpha(0f);
            searchToolbarContainer.setVisibility(View.VISIBLE);
            searchToolbarContainer.animate().translationX(0).alpha(1f).setDuration(200).setInterpolator(new DecelerateInterpolator()).start();
        } else {
            searchToolbarContainer.setTranslationX(0);
            searchToolbarContainer.setAlpha(1f);
            searchToolbarContainer.setVisibility(View.VISIBLE);
        }
    }

    @Override
    protected void initData() {
        mAdapter = new SongAdapter(searchResults);
        mAdapter.setEnableLoadMore(true);
        //初始化列表
        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        layoutManager.setOrientation(LinearLayoutManager.VERTICAL);
        resultListRcv.setLayoutManager(layoutManager);
        resultListRcv.setAdapter(mAdapter);
        mAdapter.bindToRecyclerView(resultListRcv);
        if (mPresenter != null) {
            mPresenter.getSearchHistory();
            mPresenter.getHotSearchInfo();
        }
    }

    @Override
    protected void initInjector() {
        mActivityComponent.inject(this);
    }

    @SuppressWarnings({"unchecked", "varargs"})
    @Override
    protected void listener() {
        mAdapter.setOnItemClickListener((adapter, view, position) -> {
            Music music = searchResults.get(position);
            LogUtil.e(TAG, music.toString());
            PlayManager.playOnline(music);
            NavigationHelper.INSTANCE.navigateToPlaying(this, view.findViewById(R.id.iv_cover));
        });
        searchEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                String newText = searchEditText.getText().toString();
                if (newText.length() == 0) {
                    mPresenter.getSearchHistory();
                    updateHistoryPanel(true);
                } else {
                    searchLocal(newText);
                }
            }
        });

        searchEditText.setOnEditorActionListener((v, actionId, event) -> {
            LogUtil.d(TAG, "onEditorAction() called with: v = [" + v + "], actionId = [" + actionId + "], event = [" + event + "]");
            if (event != null && (event.getKeyCode() == KeyEvent.KEYCODE_ENTER || event.getAction() == EditorInfo.IME_ACTION_SEARCH)) {
                search(searchEditText.getText().toString());
                return true;
            }
            return false;
        });

        mAdapter.setOnItemChildClickListener((adapter, view, position) -> {
            Music music = searchResults.get(position);
            BottomDialogFragment.Companion.newInstance(music, Constants.OP_ONLINE).show(this);
        });
        mAdapter.setOnLoadMoreListener(() -> resultListRcv.postDelayed(() -> {
            if (mCurrentCounter == 0) {
                //数据全部加载完毕
                mAdapter.loadMoreEnd();
            } else {
                mOffset++;
                //成功获取更多数据
                mPresenter.search(queryString, filter, limit, mOffset);
            }
        }, 1000), resultListRcv);


    }

    //
    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_search, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
            case R.id.menu_search:
                queryString = searchEditText.getText().toString().trim();
                search(queryString);
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    SearchEngine.Filter filter = SearchEngine.Filter.ANY;

    private void changeFilter(MenuItem item, SearchEngine.Filter filter) {
        this.filter = filter;
        this.filterItemCheckedId = item.getItemId();
        item.setChecked(true);

        if (!TextUtils.isEmpty(queryString)) {
            search(queryString);
        }
    }

    int filterItemCheckedId = -1;


    /**
     * 本地搜索
     *
     * @param query
     */
    private void searchLocal(String query) {
        if (query != null && query.length() > 0) {
            searchResults.clear();
            queryString = query;
            updateHistoryPanel(false);
            if (mPresenter != null) {
                mPresenter.searchLocal(query);
            }
        }
    }


    private void search(String query) {
        if (query != null && query.length() > 0) {
            mOffset = 0;
            searchResults.clear();
            queryString = query;
            updateHistoryPanel(false);
            mPresenter.saveQueryInfo(query);
            mPresenter.search(query, filter, limit, mOffset);
        }
    }


    @Override
    public void showLoading() {
        super.showLoading();
    }

    @Override
    public void hideLoading() {
        super.hideLoading();
    }

    @Override
    public void showSearchResult(List<Music> list) {
        searchResults.addAll(list);
        mAdapter.setNewData(searchResults);
        mAdapter.loadMoreComplete();
        mCurrentCounter = mAdapter.getData().size();

        if (searchResults.size() == 0) {
            showEmptyView();
        }
        LogUtil.e("search", mCurrentCounter + "--" + mCurrentCounter + "--" + mOffset);
    }

    @Override
    public void showHotSearchInfo(@NonNull List<HotSearchBean> result) {
        if (result.size() > 0) {
            hotSearchView.setVisibility(View.VISIBLE);
            AnimationUtils.animateView(hotSearchView, true, 600);
        } else hotSearchView.setVisibility(View.GONE);
        if (hotSearchAdapter == null) {
            hotSearchAdapter = new HotSearchAdapter(result);
            FlexboxLayoutManager layoutManager = new FlexboxLayoutManager(this);
            layoutManager.setFlexDirection(FlexDirection.ROW);
            layoutManager.setJustifyContent(JustifyContent.FLEX_START);
            hotSearchRcv.setLayoutManager(layoutManager);
            hotSearchRcv.setAdapter(hotSearchAdapter);
            hotSearchAdapter.bindToRecyclerView(hotSearchRcv);
            hotSearchAdapter.setOnItemClickListener((adapter, view, position) -> {
            });
            hotSearchAdapter.setOnItemChildClickListener((adapter, view, position) -> {
                if (view.getId() == R.id.titleTv) {
                    searchEditText.setText(result.get(position).getTitle());
                    searchEditText.setSelection(result.get(position).getTitle().length());
                    search(result.get(position).getTitle());
                }
            });
        } else {
            hotSearchAdapter.setNewData(result);
        }
    }

    @Override
    public void showSearchHistory(List<SearchHistoryBean> result) {
        suggestions = result;
        if (historyAdapter == null) {
            historyAdapter = new SearchHistoryAdapter(suggestions);
            historyRcv.setLayoutManager(new LinearLayoutManager(this));
            historyRcv.setAdapter(historyAdapter);
            historyAdapter.bindToRecyclerView(historyRcv);
            historyAdapter.setOnItemLongClickListener((adapter, view, position) -> {
                return false;
            });
            historyAdapter.setOnItemClickListener((adapter, view, position) -> {
            });
            historyAdapter.setOnItemChildClickListener((adapter, view, position) -> {
                if (view.getId() == R.id.history_search) {
                    searchEditText.setText(suggestions.get(position).getTitle());
                    searchEditText.setSelection(suggestions.get(position).getTitle().length());
                    search(suggestions.get(position).getTitle());
                } else if (view.getId() == R.id.deleteView) {
                    DaoLitepal.INSTANCE.deleteSearchInfo(suggestions.get(position).getTitle());
                    historyAdapter.remove(position);
                }
            });
        } else {
            suggestions = result;
            historyAdapter.setNewData(result);
        }
    }

    @Override
    public void showEmptyView() {
        mAdapter.setEmptyView(R.layout.view_song_empty);
    }

    private void updateHistoryPanel(boolean isShow) {
        if (isShow) {
            resultListRcv.setVisibility(View.GONE);
            historyPanel.setVisibility(View.VISIBLE);
        } else {
            resultListRcv.setVisibility(View.VISIBLE);
            historyPanel.setVisibility(View.GONE);
        }
    }
}

File: app/src/main/java/com/cyl/musiclake/ui/music/search/SearchActivity.kt
package com.cyl.musiclake.ui.music.search

import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.view.KeyEvent
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.animation.DecelerateInterpolator
import android.view.inputmethod.EditorInfo
import com.cyl.musiclake.R
import com.cyl.musiclake.bean.HotSearchBean
import com.cyl.musiclake.bean.Music
import com.cyl.musiclake.bean.SearchHistoryBean
import com.cyl.musiclake.data.db.DaoLitepal
import com.cyl.musiclake.common.Extras
import com.cyl.musiclake.ui.base.BaseActivity
import com.cyl.musiclake.ui.main.PageAdapter
import com.cyl.musiclake.ui.music.search.fragment.SearchSongsFragment
import com.cyl.musiclake.ui.music.search.fragment.YoutubeSearchFragment
import com.cyl.musiclake.utils.AnimationUtils
import com.cyl.musiclake.utils.Tools
import com.google.android.flexbox.FlexDirection
import com.google.android.flexbox.FlexboxLayoutManager
import com.google.android.flexbox.JustifyContent
import kotlinx.android.synthetic.main.activity_search.*
import kotlinx.android.synthetic.main.toolbar_search_layout.*
import java.util.*

/**
 * 作者：yonglong on 2016/9/15 12:32
 * 邮箱：643872807@qq.com
 * 版本：2.5
 * 描述:搜索功能
 */
class SearchActivity : BaseActivity<SearchPresenter>(), SearchContract.View {
    /**
     * 搜索信息
     */
    private var queryString: String? = null
    /**
     * 搜索结果
     */
    private val searchResults = mutableListOf<Music>()
    /**
     * 歌曲列表
     */
    private var songList = mutableListOf<Music>()

    /**
     * 适配器
     */
    private var historyAdapter: SearchHistoryAdapter? = null
    private var hotSearchAdapter: HotSearchAdapter? = null
    /**
     * 搜索历史
     */
    private var searchHistory: MutableList<SearchHistoryBean> = ArrayList()

    override fun getLayoutResID(): Int {
        return R.layout.activity_search
    }

    override fun initView() {
        showSearchAnimation()
    }

    /**
     * 显示搜索弹出动画
     */
    private fun showSearchAnimation() {
        searchEditText.setText(queryString)
        if (TextUtils.isEmpty(queryString) || TextUtils.isEmpty(searchEditText.text)) {
            searchToolbarContainer.translationX = 100f
            searchToolbarContainer.alpha = 0f
            searchToolbarContainer.visibility = View.VISIBLE
            searchToolbarContainer.animate().translationX(0f).alpha(1f).setDuration(200).setInterpolator(DecelerateInterpolator()).start()
        } else {
            searchToolbarContainer.translationX = 0f
            searchToolbarContainer.alpha = 1f
            searchToolbarContainer.visibility = View.VISIBLE
        }
    }

    override fun initData() {
        //初始化列表
//        val layoutManager = LinearLayoutManager(this)
//        layoutManager.orientation = LinearLayoutManager.VERTICAL
//        resultListRcv.layoutManager = layoutManager
//        resultListRcv.adapter = mAdapter
//        mAdapter.bindToRecyclerView(resultListRcv)

        //获取搜索历史
        mPresenter?.getSearchHistory()

        if (!intent.getBooleanExtra("is_playlist", false)) {
            //获取热搜
            mPresenter?.getHotSearchInfo()
        } else {
        }

        //传值搜索
        if (intent.getStringExtra(Extras.SEARCH_INFO)?.isNotEmpty() == true) {
            searchEditText.setText(intent.getStringExtra(Extras.SEARCH_INFO))
        }
    }

    override fun initInjector() {
        mActivityComponent.inject(this)
    }

    /**
     * 监听事件
     */
    override fun listener() {
        clearAllIv.setOnClickListener {
            DaoLitepal.clearAllSearch()
            searchHistory.clear()
            historyAdapter?.setNewData(searchHistory)
        }
        clearSearchIv.setOnClickListener {
            queryString = ""
            searchEditText.setText("")
            clearSearchIv.visibility = View.GONE
        }
        searchEditText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}

            override fun afterTextChanged(s: Editable) {
                val newText = searchEditText.text.toString()
                clearSearchIv.visibility = View.VISIBLE
                if (newText.isEmpty()) {
                    mPresenter?.getSearchHistory()
                    updateHistoryPanel(true)
                    clearSearchIv.visibility = View.GONE
                }
            }
        })

        searchEditText.setOnEditorActionListener { _, _, event ->
            if (event != null && (event.keyCode == KeyEvent.KEYCODE_ENTER || event.action == EditorInfo.IME_ACTION_SEARCH)) {
                search(searchEditText.text.toString())
                return@setOnEditorActionListener true
            }
            false
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_search, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> finish()
            R.id.menu_search -> {
                queryString = searchEditText.text.toString().trim { it <= ' ' }
                search(queryString)
            }
        }
        return true
    }

    /**
     * 本地搜索
     *
     * @param query
     */
    private fun searchLocal(query: String?) {
        if (query != null && query.isNotEmpty()) {
            searchResults.clear()
            queryString = query
            updateHistoryPanel(false)
            mPresenter?.searchLocal(query)
        }
    }

    /**
     * 在线搜索
     * @param query 搜索内容
     */
    private fun search(query: String?) {
        if (query != null && query.isNotEmpty()) {
            showLoading()
            searchResults.clear()
            queryString = query
            searchEditText.clearFocus()
            Tools.hideInputView(searchEditText)
            updateHistoryPanel(false)
            mPresenter?.saveQueryInfo(query)
            setupViewPager()
//            mPresenter?.search(query, SearchEngine.Filter.ANY, limit, mOffset)
        }
    }

    /**
     * 显示搜索记录
     */
    override fun showSearchResult(list: MutableList<Music>) {

    }

    /**
     * 设置热搜
     */
    override fun showHotSearchInfo(result: MutableList<HotSearchBean>) {
        if (result.size > 0) {
            hotSearchView.visibility = View.VISIBLE
            AnimationUtils.animateView(hotSearchView, true, 600)
        } else
            hotSearchView.visibility = View.GONE
        if (hotSearchAdapter == null) {
            hotSearchAdapter = HotSearchAdapter(result)
            val layoutManager = FlexboxLayoutManager(this)
            layoutManager.flexDirection = FlexDirection.ROW
            layoutManager.justifyContent = JustifyContent.FLEX_START
            hotSearchRcv.layoutManager = layoutManager
            hotSearchRcv.adapter = hotSearchAdapter
            hotSearchAdapter?.setOnItemClickListener { _, _, _ -> }
            hotSearchAdapter?.setOnItemChildClickListener { _, view, position ->
                if (view.id == R.id.titleTv) {
                    searchEditText.setText(result[position].title)
                    searchEditText.setSelection(result[position].title?.length ?: 0)
                    search(result[position].title)
                }
            }
        } else {
            hotSearchAdapter?.setNewData(result)
        }
    }

    /**
     * 显示历史
     */
    override fun showSearchHistory(result: MutableList<SearchHistoryBean>) {
        searchHistory = result
        if (historyAdapter == null) {
            historyAdapter = SearchHistoryAdapter(searchHistory)
            historyRcv.layoutManager = androidx.recyclerview.widget.LinearLayoutManager(this)
            historyRcv.adapter = historyAdapter
            historyAdapter?.setOnItemLongClickListener { _, _, _ -> false }
            historyAdapter?.setOnItemClickListener { _, _, _ -> }
            historyAdapter?.setOnItemChildClickListener { _, view, position ->
                if (view.id == R.id.history_search) {
                    searchEditText.setText(searchHistory[position].title)
                    searchEditText.setSelection(searchHistory[position].title?.length ?: 0)
                    search(searchHistory[position].title)
                } else if (view.id == R.id.deleteView) {
                    searchHistory[position].title?.let { DaoLitepal.deleteSearchInfo(it) }
                    historyAdapter?.remove(position)
                }
            }
        } else {
            searchHistory = result
            historyAdapter?.setNewData(result)
        }
    }

    override fun onResume() {
        super.onResume()
        //强制隐藏键盘输入法
        Tools.hideInputView(searchEditText)
    }

    /**
     * 是否显示搜索历史列表
     * @param isShow
     */
    private fun updateHistoryPanel(isShow: Boolean) {
        if (isShow) {
            tabs?.visibility = View.GONE
            viewPager?.visibility = View.GONE
            historyPanel.visibility = View.VISIBLE
        } else {
            historyPanel.visibility = View.GONE
            tabs?.visibility = View.VISIBLE
            viewPager?.visibility = View.VISIBLE
            tabs?.setupWithViewPager(viewPager)
            viewPager.currentItem = 0
        }
    }

    /**
     * 设置搜索结果列表
     */
    private fun setupViewPager() {
        val mAdapter = PageAdapter(supportFragmentManager)
        mAdapter.addFragment(SearchSongsFragment.newInstance(queryString, SearchEngine.Filter.QQ), "QQ")
        mAdapter.addFragment(SearchSongsFragment.newInstance(queryString, SearchEngine.Filter.NETEASE), "网易云")
        mAdapter.addFragment(SearchSongsFragment.newInstance(queryString, SearchEngine.Filter.XIAMI), "虾米")
        mAdapter.addFragment(SearchSongsFragment.newInstance(queryString, SearchEngine.Filter.BAIDU), "百度")
        mAdapter.addFragment(YoutubeSearchFragment.newInstance(queryString), "Youtube")
        viewPager?.adapter = mAdapter
        viewPager?.offscreenPageLimit = 4
        hideLoading()
    }
}
